sqlmesh/core/engine_adapter/snowflake.py

```diff
@@ -326,12 +345,44 @@ def _get_data_objects(
         ]
 
     def set_current_catalog(self, catalog: str) -> None:
-        self.execute(exp.Use(this=exp.to_identifier(catalog)))
+        self.execute(exp.Use(this=normalize_identifiers(catalog, dialect=self.dialect)))
+
+    def set_current_schema(self, schema: str) -> None:
+        self.execute(exp.Use(kind="SCHEMA", this=to_schema(schema)))
+
+    def _to_sql(self, expression: exp.Expression, quote: bool = True, **kwargs: t.Any) -> str:
+        # note: important to use self._default_catalog instead of the self.default_catalog property
+        # otherwise we get RecursionError: maximum recursion depth exceeded
+        # because it calls get_current_catalog(), which executes a query, which needs the default catalog, which calls get_current_catalog()... etc
+        if self._default_catalog:
+
+            def unquote_and_lower(identifier: str) -> str:
+                return exp.parse_identifier(identifier).name.lower()
```

Comment:
why are we doing this?

Reply:
to "fuzzy match" the catalog set on the query vs the default catalog set in the config. The idea is to distinguish catalogs the user has deliberately set vs when the default catalog has been used.

For example, let's say the source dialect is `duckdb` and the target is `snowflake`. The query is normalized in the source dialect and might contain a reference to `"erin"."schema"."table"`. However, my Snowflake gateway is configured with `database: ERIN`, so this is wrong, we need to change the reference to `ERIN."schema"."table"` or ideally `"ERIN"."schema"."table"` to be truly unambiguous,

Since `ERIN` is not the same as `"erin"`, I cant do a direct string comparison. But if I remove the quotes and lowercase both sides, `erin`==`erin` so I can assume that the default catalog is in use and can replace the catalog reference in the query with the _normalized_ version from the gateway config.

If the user has specified some other catalog manually, eg the model fqn is `"external"."schema"."table"` then it wont match because `erin` != `external` so the catalog is not touched.

Doing it like this means it'll also work if I do something crazy like use a case sensitive catalog in my connection config, eg `database: '"EriN"'`.

Since `unquote_and_lower('"ErIN"') == unquote_and_lower('"erin"')` I know to replace the catalog

Reply:
perfect. can we add a comment or change the name such that the intention of the fuzzy match is clear?